README.md
[![Docker Image CI Main](https://github.com/GamingBunker/AnimePlex/actions/workflows/docker-image.yml/badge.svg?branch=main)](https://github.com/GamingBunker/AnimePlex/actions/workflows/docker-image.yml) 

# Anime Plex 🎬
Scarica le serie tv/film dal sito [AnimeSaturn](https://www.animesaturn.it/) e mette nella cartella Plex.

## La struttura del progetto
Il progetto si suddivide in 9 progetti:
- 🧮Api Service (C#)
- 📩Download Service (C#)
- 📨Notify Service (C#)
- 💾Update Service (C#)
- 💽Upgrade Service (C#)
- 💱Conversion Service (C#)
- 🏠Room server (Hapi)
- 📁Path server (Nodejs)
- 🌐Web Server([Nuxtjs](https://nuxtjs.org/))

Servizi utilizzati:
- 🐰[RabbitMQ](https://www.rabbitmq.com/)
- 🗄[Postgresql](https://www.postgresql.org/)

### Lista di immagini disponibili su DockerHub
| Nome Immagine | Link |
| ------ | ------ |
| 🧮Api Service | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-api) |
| 📩Download Service | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-downloadservice) |
| 📨Notify Service | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-notifyservice) |
| 💾Update Service | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-updateservice) |
| 💽Upgrade Service | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-upgradeservice) |
| 💱Conversion Service | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-conversionservice) |
| 🏠Room server (Hapi) | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-roomserver) |
| 🌐Web Client | [Link](https://hub.docker.com/r/kju7pwd2/animeplex-web) |

## 🌐Web Server
Questo progetto verrà utilizzato per gli utenti che vorranno visualizzare e scaricare gli episodi.

### Expose Ports:
- 3000 tcp

### Variabili globali richiesti:
```sh
example:
    #--- API ---
    API_BASE_URL: 'http://localhost:3333' #http://localhost:5000 [default]

    #--- General ---
    NODE_TLS_REJECT_UNAUTHORIZED: 0

    #--- Path ---
    HTTP_PATH: 'http://localhost:3333' #http://localhost:5002 [default]
    BASE_PATH: "/path" #"/public" [default]

    #--- WebSocket ---
    SOCKET_PATH: "ws://localhost:1111/path" #ws://localhost:5001/room [default]
    
    #--- Share link ---
    SHARE_ROOM: "http://localhost:33333" #http://localhost:3000 [default]
```

## 🧮Api Service
Questo progetto verrà utilizzato per esporre i dati in maniera facile e veloce con il database postgresql.

### Expose Ports:
- 80 tcp

### Information general:
- `not` require volume mounted on Docker
### Variabili globali richiesti:
```sh
example:
    #--- DB ---
    DATABASE_CONNECTION: User ID=guest;Password=guest;Host=localhost;Port=33333;Database=db; [require]
    
    #--- Rabbit ---
    USERNAME_RABBIT: "guest" #guest [default]
    PASSWORD_RABBIT: "guest" #guest [default]
    ADDRESS_RABBIT: "localhost" #localhost [default]

    #--- API ---
    PORT_API: "33333" #5000 [default]
    
    #--- Logger ---
    LOG_LEVEL: "Debug|Info|Error" #Info [default]
    WEBHOOK_DISCORD_DEBUG: "url" [not require]
    
    #--- General ---
    BASE_PATH: "/folder/anime" or "D:\\\\Directory\Anime" #/ [default]
    LIMIT_THREAD_PARALLEL: "8" #5 [default]
```

## 💾Update Service
Questo progetto verrà utilizzato per controllare se sono presenti nel file locale se non ci sono, invia un messaggio a DownloadService che scarica l'episodio mancante.
### Information general:
- `require` volume mounted on Docker
### Variabili globali richiesti:
```sh
example:
    #--- Rabbit ---
    USERNAME_RABBIT: "guest" #guest [default]
    PASSWORD_RABBIT: "guest" #guest [default]
    ADDRESS_RABBIT: "localhost" #localhost [default]
    
    #--- API ---
    ADDRESS_API: "localhost" #localhost [default]
    PORT_API: "33333" #5000 [default]
    PROTOCOL_API: "http" or "https" #http [default]
    
    #--- Logger ---
    LOG_LEVEL: "Debug|Info|Error" #Info [default]
    WEBHOOK_DISCORD_DEBUG: "url" [not require]
    
    #--- General ---
    BASE_PATH: "/folder/anime" or "D:\\\\Directory\Anime" #/ [default]
    TIME_REFRESH: "60000" <-- milliseconds #120000 [default] 2 minutes
    LIMIT_THREAD_PARALLEL: "8" #5 [default]
    SELECT_SERVICE: "manga or anime" #anime
```

## 💽Upgrade Service
Questo progetto verrà utilizzato per scaricare i nuovi episodi
### Information general:
- `not` require volume mounted on Docker
### Variabili globali richiesti:
```sh
example:
    #--- rabbit ---
    USERNAME_RABBIT: "guest" #guest [default]
    PASSWORD_RABBIT: "guest" #guest [default]
    ADDRESS_RABBIT: "localhost" #localhost [default]
    
    #--- API ---
    ADDRESS_API: "localhost" #localhost [default]
    PORT_API: "33333" #5000 [default]
    PROTOCOL_API: "http" or "https" #http [default]
    
    #--- Logger ---
    LOG_LEVEL: "Debug|Info|Error" #Info [default]
    WEBHOOK_DISCORD_DEBUG: "url" [not require]
    
    #--- General ---
    BASE_PATH: "/folder/anime" or "D:\\\\Directory\Anime" #http [default]
    TIME_REFRESH: "60000" <-- milliseconds #1200000 [default] 20 minutes
    LIMIT_THREAD_PARALLEL: "8" #5 [default]
    SELECT_SERVICE: "manga or anime" #anime
```

## 📩Download Service
Questo progetto verrà utilizzato per scaricare i video e mettere nella cartella.
### Information general:
- `require` volume mounted on Docker
### Variabili globali richiesti:
```sh
example:
    #--- rabbit ---
    USERNAME_RABBIT: "guest" #guest [default]
    PASSWORD_RABBIT: "guest" #guest [default]
    ADDRESS_RABBIT: "localhost" #localhost [default]
    LIMIT_CONSUMER_RABBIT: "5" #3 [default]
    
    #--- API ---
    ADDRESS_API: "localhost" #localhost [default]
    PORT_API: "33333" #3000 [default]
    PROTOCOL_API: "http" or "https" #http [default]
    
    #--- Logger ---
    LOG_LEVEL: "Debug|Info|Error" #Info [default]
    WEBHOOK_DISCORD_DEBUG: "url" [not require]
    
    #--- General ---
    LIMIT_THREAD_PARALLEL: "500" #5 [default]
    PATH_TEMP: "/tmp/folder" #D:\\TestAnime\\temp [default]
    BASE_PATH: "/folder/anime" or "D:\\\\Directory\Anime" #/ [default]
```

## 📨Notify Service
### Information general:
- `not` require volume mounted on Docker
```sh
example:
    #--- rabbit ---
    USERNAME_RABBIT: "guest" #guest [default]
    PASSWORD_RABBIT: "guest" #guest [default]
    ADDRESS_RABBIT: "localhost" #localhost [default]

    #--- API ---
    ADDRESS_API: "localhost" #localhost [default]
    PORT_API: "33333" #5000 [default]
    PROTOCOL_API: "http" or "https" #http [default]
    
    #---Webhook---
    WEBHOOK_DISCORD: "url" [require]
    
    #---logger---
    LOG_LEVEL: "Debug|Info|Error" #Info [default]
    WEBHOOK_DISCORD_DEBUG: "url" [not require]
```

## 💱Conversion Service
Questo progetto verrà utilizzato per convertire file ts in mp4 da poter riprodurre in streaming
### Information general:
- `require` volume mounted on Docker
### Variabili globali richiesti:
```sh
example:
    #--- rabbit ---
    USERNAME_RABBIT: "guest" #guest [default]
    PASSWORD_RABBIT: "guest" #guest [default]
    ADDRESS_RABBIT: "localhost" #localhost [default]
    LIMIT_CONSUMER_RABBIT: "5" #3 [default]
    
    #--- API ---
    ADDRESS_API: "localhost" #localhost [default]
    PORT_API: "33333" #3000 [default]
    PROTOCOL_API: "http" or "https" #http [default]
    
    #--- Logger ---
    LOG_LEVEL: "Debug|Info|Error" #Info [default]
    WEBHOOK_DISCORD_DEBUG: "url" [not require]
    
    #--- General ---
    PATH_TEMP: "/folder/temp" [require]
    PATH_FFMPEG: "/folder/bin" #/usr/local/bin/ffmpeg [default]
    BASE_PATH: "/folder/anime" or "D:\\\\Directory\Anime" #/ [default]
```

## 🏠Room server (Hapi)
questo progetto viene gestito le sessioni di streaming e le interazioni dei video degli altri, per esempio se viene messo in pausa tutte le persone che sono presenti in quella stanza viene messo in pausa il video.
### Information general:
- `not` require volume mounted on Docker
### Variabili globali richiesti:
```sh
example:
    #--- General ---
    HOST: "localhost" #0.0.0.0 [default]
    PORT: "33333" #1234 [default]
    PATH_URL: "/path" #/room [default]
```
## 📁Path server (Nodejs)
Questo servizio serve ad esporre i file video per web.
### Information general:
Creare un container che contiene una lts di linux.
Installare `nodejs` e `npm`, infine installare il pacchetto ftp: `npm install --global http-server`
Avviare con `http-server '/root/anime'` come avvio della macchina

Oppure si può usare questa immagine: `danjellz/http-server`, la cartella che viene esposta è la seguente: `/public`

changelog
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Released]
## [2.1.0] - 11-12-2022
### Changed
- Refactory web (change GUI and upgrade nuxt2 to nuxt3)

## [2.0.0] - 08-08-2022
### Added
- room virtual for watch video with friends
- ConversionSerivce for convert file ts to mp4 for streaming
- Websocket for change information for streaming
- Http-server for expose files video for streaming
- docker-compose-unstable (get image beta)

### Fixed
- resolve bug for get url mp4 by html

### Changed
- file docker-compose

## [1.0.1] - 20-07-2022
### Fixed
- resolve size button of mangaworld for mobile
- resolve crash upgradeService
- change domain for anime

### Changed
- update interfaces for back end

## [1.0.0] - 23-05-2022
### Added
- now support Manga (use mangaworld)

### Changed
- Banner
- Structure DB

### Fixed
- Problem refresh when delete anime/manga
- Disable autoset season for anime

## [0.3.2] - 21-04-2022
### Fixed
- order list when add new episodes how notify
- resolve problem performance of the cpu always at 100%

## [0.3.1] - 20-04-2022
### Changed
- Now use nuget of Discord Webhook

### Add
- Support parallel now with UpdateService
- Now for notify add new Anime/Episode is incluse with image 

### Fixed
- Clear code
- Resolve order list of the episodes in web
- Improve parallel
- Improve interface web general

## [0.3.0] - 17-04-2022
### Added
- Now you can delete anime (only in db!)
- Now recive also notify of when add new anime and remove old anime

### Changed
- change color for download anime (now is yellow)

### Fixed
- Refactory code c#
- Problem check already downloaded anime with bar search to animeSaturn

### Bug
- Resolve problem method Hash when open file busy
- DownloadService delete messages duplicated

## [0.2.5] - 10-04-2022
### Fixed
- Resolve problem show disk space

## [0.2.4] - 10-04-2022
### Fixed
- Resolve problem get urls for episodes during download in parallel
- UpdateService also update the renaming of a file

### Add
- Show information space disk
- Show information health of services
- Add slide bar for see informations

## [0.2.3] - 08-04-2022
### Add
- Webhook Discord for log

### Fixed
- Module HtmlAnimeSaturn now work in parallel for download urls of episodes
- Module Download now work in parallel for second method (use with buffers) of episodes
- Fixed overflow memory when calculate hash of one episode over 1GB size

## [0.2.2] - 06-04-2022
### Add
- Setup default environment global

### Fixed
- Change structure for logger

## [0.2.1] - 05-04-2022
### Add
- Notify service that send messages to Discord with Webhook

### Fixed
- Add environment logger for reduced lines of log

## [0.2.0] - 01-04-2022
### Add
- Tracking files and movements (modified class models and added new table)
- Check Hash256
- Browser support now Mobile
- Add all description of an anime into card, if anime not already downloaded show a link for redirect original page with information

### Fixed
- Improve call API
- Improve conversion class model to class dto or reverse
- If Anime already downloaded not show button for download episodes

### Changed
- Split two part of UpdateService, UpdateService check integry files on local and UpgradeService check new or missing episodes into db
- Buttons Search
- Size pagination
- replace 'previous' and 'next' with arrows in pagination
- Show status of downloads after press a button


## [0.1.2] - 29-03-2022
### Add
- Add button re-download anime if download are stop

### Fixed
- when download is filed, bar progress is red with write 'filed' instead of percentual with value 100%
- when download is completed, bar progress is green with write 'completed' instead of percentual
- Resolve problem open file

### Changed
- DownloadService start timeout 0 to max limit defined and print log if error download with attempts remains, if download is success reset timeout

## [0.1.1] - 29-03-2022
### Added
- new button into bar search for get all anime

### Changed
- color buttons into bar search


## [0.1.0] - 27-03-2022
### Version Stable

src/web/nuxt.config.js
export default {
  // Global page headers: https://go.nuxtjs.dev/config-head
  head: {
    title: 'Anime Plex',
    htmlAttrs: {
      lang: 'en'
    },
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      { hid: 'description', name: 'description', content: '' },
      { name: 'format-detection', content: 'telephone=no' }
    ],
    link: [
      { rel:"stylesheet", href:'https://cdn.jsdelivr.net/npm/bootstrap-icons@1.8.1/font/bootstrap-icons.css' },
      { rel:"stylesheet", href:'https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/css/bootstrap.min.css', integrity:'sha384-1BmE4kWBq78iYhFldvKuhfTAU6auU8tT94WrHftjDbrCEXSU1oBoqyl2QvZ6jIW3', crossorigin: 'anonymous'},
      { rel:"stylesheet", href:"https://vjs.zencdn.net/7.20.1/video-js.css" }
    ],
    script:[
      { src:"https://vjs.zencdn.net/7.20.1/video.min.js" }
    ]
  },

  // Global CSS: https://go.nuxtjs.dev/config-css
  css: [
  ],

  // Plugins to run before rendering page: https://go.nuxtjs.dev/config-plugins
  plugins: [
    '@/plugins/axios',
  ],

  // Auto import components: https://go.nuxtjs.dev/config-components
  components: true,

  // Modules for dev and build (recommended): https://go.nuxtjs.dev/config-modules
  buildModules: [
  ],

  // Modules: https://go.nuxtjs.dev/config-modules
  modules: [
    // https://go.nuxtjs.dev/axios
    '@nuxtjs/axios',
    'bootstrap-vue/nuxt'
  ],
  
  // Axios module configuration: https://go.nuxtjs.dev/config-axios
  axios: {
    // Workaround to avoid enforcing hard-coded localhost:3000: https://github.com/nuxt-community/axios-module/issues/308
    baseURL: '/',
  },

  // Build Configuration: https://go.nuxtjs.dev/config-build
  build: {
  },

  server:{
    host: "0.0.0.0"
  },

  publicRuntimeConfig: {
    ipAPI: process.env.HOST_API || "localhost",
    portAPI: process.env.PORT_API || "5000",
    protocolAPI: process.env.PROTOCOL_API || "http",

    ipWebSocket: process.env.HOST_WS || "ws://localhost:1234/room",
    basePath: process.env.BASE_PATH || "/",

    ipHttpServer: process.env.HOST_HTTP_SERVER || "localhost",
    portHttpServer: process.env.PORT_HTTP_SERVER || "8080",

    shareRoom: process.env.SHARE_ROOM || 'localhost:3000'
  }
}
